### PR TITLE
Add Amazon SES Support

### PR DESCRIPTION
**Pull request by wreality**
No guarantees this doesn't need more work / an appropriate level of skepticism, but this PR adds sending and webhook support for Amazon SES.

### CODE REVIEW

**Comment by maxgurewitz** [comment]
wow thank you @wreality!

File: packages/api/src/controllers/webhooksController.ts [hunk 1]
```diff
@@ -115,6 +126,59 @@ export default async function webhookController(fastify: FastifyInstance) {
     },
   );
 
+  fastify.withTypeProvider<TypeBoxTypeProvider>().post(
+    "/amazonses",
```
Comment (maxgurewitz) [inline packages/api/src/controllers/webhooksController.ts:130]: Small nitpick, but can we rename this `amazon-ses`?

File: packages/api/src/controllers/webhooksController.ts [hunk 2]
```diff
@@ -115,6 +126,59 @@ export default async function webhookController(fastify: FastifyInstance) {
...
+    {
+      schema: {
+        description: "Used to consume amazonses notification events.",
+        tags: ["Webhooks"],
+        body: AmazonSNSEvent,
+      },
+      onRequest: (req, _, done) => {
```
Comment (maxgurewitz) [inline packages/api/src/controllers/webhooksController.ts:137]: Why is this necessary?

File: packages/api/src/controllers/webhooksController.ts [hunk 3]
```diff
@@ -115,6 +126,59 @@ export default async function webhookController(fastify: FastifyInstance) {
...
+      // Validate the signature
+      const valid = await validSNSSignature(request.body);
+
+      if (valid.isErr()) {
+        logger().error("Invalid signature for AmazonSES webhook.", valid.error);
+        return reply.status(401).send({ message: "Invalid signature" });
+      }
+
+      switch (request.body.Type) {
+        case AmazonSNSEventTypes.SubscriptionConfirmation:
```
Comment (maxgurewitz) [inline packages/api/src/controllers/webhooksController.ts:155]: The effect of this code is to have the `SubscriptionConfirmation` and `UnsubscribeConfirmation` be handled by the same code block. Is that the intention?

File: packages/api/src/controllers/webhooksController.ts [hunk 4]
```diff
@@ -115,6 +126,59 @@ export default async function webhookController(fastify: FastifyInstance) {
...
+        case AmazonSNSEventTypes.UnsubscribeConfirmation:
```
Comment (maxgurewitz) [inline packages/api/src/controllers/webhooksController.ts:156]: Subscriptions are handled internally, within dittofeed. In the case of other email service providers, like sendgrid, we ignore subscription change events, because we disable the subscription functionality within sendgrid and delegate it to Dittofeed. We should likely do the same for SES. Does that make sense?

File: packages/backend-lib/src/destinations/amazonses.ts [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+  InternalEventType,
+} from "../types";
+
+function unwrapTag(tagName: string, tags: Record<string, string[]>) {
+  if (!tags[tagName]) {
+    return null;
+  }
+  // disable the nonnull assertion eslint rule
+  // eslint-disable-next-line
+  const tag = tags[tagName]!;
```
Comment (maxgurewitz) [inline packages/backend-lib/src/destinations/amazonses.ts:35]: rather than asserting non-null we can do something like `return tags[tagName]?[0] ?? null`

File: packages/backend-lib/src/destinations/amazonses.ts [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+  );
+}
+
+export async function submitAmazonSesEvents(
+  event: AmazonSesEventPayload,
+): Promise<ResultAsync<void, Error>> {
+  // TODO: Amazon may batch requests (if we send with multiple To: addresses? or with the BatchTemplated endpoint).  We should map over the receipients.
+  logger().debug(event);
+
+  const workspaceId = unwrapTag("workspaceId", event.mail.tags) ?? null;
```
Comment (maxgurewitz) [inline packages/backend-lib/src/destinations/amazonses.ts:118]: Is it necessary to do `?? null` it looks like this function returns `string | null` already.

File: packages/backend-lib/src/types.ts [hunk 7]
```diff
@@ -146,6 +146,238 @@ export const DecodedJwt = Type.Object({
...
+    timestamp: Type.String(),
+    feedbackId: Type.String(),
+    complaintSubType: Type.Enum(AmazonSesComplaintSubType),
+    userAgent: Type.Optional(Type.String()),
+    complaintFeedbackType: Type.Optional(Type.String()),
+    arrivalDate: Type.Optional(Type.String()),
+  }),
+});
+
+export const AmazonSesDeliveryEvent = Type.Object({
```
Comment (maxgurewitz) [inline packages/backend-lib/src/types.ts:292]: A lot of thorough typing here thank you!

File: packages/dashboard/src/pages/settings.page.tsx [hunk 8]
```diff
@@ -665,6 +666,73 @@ function SendGridConfig() {
   );
 }
 
+function AmazonSesConfig() {
+  const { secretAvailability } = useAppStorePick(["secretAvailability"]);
+
+  return (
+    <Fields
+      sections={[
```
Comment (maxgurewitz) [inline packages/dashboard/src/pages/settings.page.tsx:674]: 👏

File: packages/api/src/controllers/webhooksController.ts (hunk 1 above)
Comment (wreality) [reply packages/api/src/controllers/webhooksController.ts:130]: Yep, it could also be `aws` if we'd like to avoid multiple words.

File: packages/api/src/controllers/webhooksController.ts (hunk 2 above)
Comment (wreality) [reply packages/api/src/controllers/webhooksController.ts:137]: Amazon sends webhooks with a JSON body and a content type of "text/plain", annoyingly.  We could just JSON.parse the body in the handler, but this way the schema checking can still do its thing in fastify.

File: packages/api/src/controllers/webhooksController.ts (hunk 3 above)
Comment (wreality) [reply packages/api/src/controllers/webhooksController.ts:155]: Yes, indeed.  I can refactor if you'd prefer not use a fall through case. (Its also a three case switch.  It would be just as clean as if/else.  Your call ofc.

File: packages/api/src/controllers/webhooksController.ts (hunk 4 above)
Comment (wreality) [reply packages/api/src/controllers/webhooksController.ts:156]: It does, but this is about the notification service (unrelated to SES) confirming the webhook is alive.  When the destination is added in the AWS console it sends a "SubscriptionConfirmation" to the webhook.  The webhook has to perform a GET on the "SubscribeUrl" property that was sent with the confirmation in order for the destination to become active and start receiving events.

File: packages/backend-lib/src/destinations/amazonses.ts (hunk 5 above)
Comment (wreality) [reply packages/backend-lib/src/destinations/amazonses.ts:35]: Yep, for some reason TS was being grumpy about that syntax last night... I'll take another look and see what's going on.

File: packages/backend-lib/src/types.ts (hunk 7 above)
Comment (wreality) [reply packages/backend-lib/src/types.ts:292]: I was hoping it wasn't too thorough :-)

**Comment by wreality** [comment]
@maxgurewitz Awesome, I'll make the changes and update later tonight.  There a couple of questions in my responses if you get a chance, but otherwise thanks for the review!!

File: packages/api/src/controllers/webhooksController.ts (hunk 2 above)
Comment (maxgurewitz) [reply packages/api/src/controllers/webhooksController.ts:137]: That's annoying! Can you leave a short comment to that effect?

File: packages/api/src/controllers/webhooksController.ts (hunk 1 above)
Comment (maxgurewitz) [reply packages/api/src/controllers/webhooksController.ts:130]: Either `aws-ses`, `ses` or `amazon-ses` work for me!

File: packages/api/src/controllers/webhooksController.ts (hunk 4 above)
Comment (maxgurewitz) [reply packages/api/src/controllers/webhooksController.ts:156]: Ah ok, thank you for the clarification! Can you write a short comment?

File: packages/api/src/controllers/webhooksController.ts (hunk 3 above)
Comment (maxgurewitz) [reply packages/api/src/controllers/webhooksController.ts:155]: This is fine, just write a short comment please! Philosophically opposed to bikeshedding code style, as long as the intent is clear.

File: packages/backend-lib/src/destinations/amazonses.ts (hunk 5 above)
Comment (wreality) [reply packages/backend-lib/src/destinations/amazonses.ts:35]: This is totally what was tripping me up, I thought the syntax was what you suggested as well.  Turns out, its actually `tags[tagName]?.[0]`.  Grr..

Comment (maxgurewitz) [reply packages/backend-lib/src/destinations/amazonses.ts]: Perfect! That syntax is counterintuitive which must be why i forgot it!

**Comment by maxgurewitz** [comment]
Cool! I'm going to try to run through these instructions to QA with our SES account, and I'll get back to you!

File: packages/docs/integrations/channels/amazonses.mdx [hunk 9]
```diff
@@ -0,0 +1,196 @@
...
+Select HTTPS for "Protocol"
+
+Enter `https://[mydittofeed]/api/webhooks/amazon-ses` replacing `[mydittofeed]` with the domain name of your dittofeed instance.
+
+Click "Create subscription".
+
+The subscription should be confirmed almost immediately.  If not, check that the URL was entered correctly and your instance is reachable from the open internet.
+
+Send a test email and check that new events populate in the events page.
+
```
Comment (maxgurewitz) [inline packages/docs/integrations/channels/amazonses.mdx:194]: Could add a sentence here about finding and visiting the confirmation url in the API logs?

**Review by maxgurewitz — CHANGES_REQUESTED** [review]
Love the very thorough docs! Especially appreciate the mention of best practices. Have some feedback on the docs. Let me know what you think!

**Review by maxgurewitz — APPROVED** [review]
Aamazing @wreality can't thank you enough this is great work!
